llarp/router/outbound_message_handler.cpp

```diff
@@ -0,0 +1,217 @@
+
+#include <messages/link_message.hpp>
+#include <router/i_outbound_session_maker.hpp>
+#include <link/i_link_manager.hpp>
+#include <constants/link_layer.hpp>
+#include <util/memfn.hpp>
+#include <util/status.hpp>
+
+namespace llarp
+{
+
+  bool
+  OutboundMessageHandler::QueueMessage(const RouterID &remote,
+                                       const ILinkMessage *msg,
+                                       SendStatusHandler callback)
+  {
+    std::array< byte_t, MAX_LINK_MSG_SIZE > linkmsg_buffer;
+    llarp_buffer_t buf(linkmsg_buffer);
+
+    if (!EncodeBuffer(msg, buf))
+    {
+      return false;
+    }
+
+    buf.sz = buf.cur - buf.base;
+
+    Message message{buf.sz, callback};
+
+    memcpy(message.first.data(), buf.base, buf.sz);
+
+    if (SendIfSession(remote, message))
+    {
+      return true;
+    }
+
+    bool shouldCreateSession = false;
+    {
+      util::Lock l(&_mutex);
+
+      // create queue for <remote> if it doesn't exist, and get iterator
+      auto itr_pair = outboundMessageQueue.emplace(remote, MessageQueue());
+
+      itr_pair.first->second.push(std::move(message));
+
+      shouldCreateSession = itr_pair.second;
+    }
+
+    if (shouldCreateSession)
+    {
+      QueueSessionCreation(remote);
+    }
+
+    return true;
+  }
+
+  //TODO: this
+  util::StatusObject
+  OutboundMessageHandler::ExtractStatus() const
+  {
+    util::StatusObject status{};
+    return status;
+  }
+
+  void
+  OutboundMessageHandler::Init(ILinkManager* linkManager,
+                               std::shared_ptr< Logic > logic)
+  {
+    _linkManager = linkManager;
+    _logic = logic;
+  }
+
+  void
+  OutboundMessageHandler::OnSessionEstablished(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Success);
+  }
+
+  void
+  OutboundMessageHandler::OnConnectFailure(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Timeout);
+  }
+
+  void
+  OutboundMessageHandler::OnRouterNotFound(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::RouterNotFound);
+  }
+
+  void
+  OutboundMessageHandler::OnInvalidRouter(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::InvalidRouter);
+  }
+
+  void
+  OutboundMessageHandler::OnNoLink(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::NoLink);
+  }
+
+  void
+  OutboundMessageHandler::OnSessionResult(const RouterID &router,
+                                          const SessionResult result)
+  {
+    switch(result)
+    {
+      case SessionResult::Establish:
+        OnSessionEstablished(router);
+        break;
+      case SessionResult::Timeout:
+        OnConnectFailure(router);
+        break;
+      case SessionResult::RouterNotFound:
+        OnRouterNotFound(router);
+        break;
+      case SessionResult::InvalidRouter:
+        OnInvalidRouter(router);
+        break;
+      case SessionResult::NoLink:
+        OnNoLink(router);
+        break;
+      default:
```

Comment:
this  should probably error on default case "just in case"

Reply:
I guess, but that would be a very interesting case.

Reply:
Added `LogError` and abort.